extensions/test/BUILD.gn

```diff
@@ -0,0 +1,157 @@
+    "$root_gen_dir/xwalk/extensions/test/test.h",
+    "bad_extension_test.cc",
+    "conflicting_entry_points.cc",
+    "context_destruction.cc",
+    "crash_extension_process.cc",
+    "export_object.cc",
+    "extension_in_iframe.cc",
+    "external_extension.cc",
+    "external_extension_multi_process.cc",
+    "in_process_threads_browsertest.cc",
+    "internal_extension_browsertest.cc",
+    "internal_extension_browsertest.h",
+    "namespace_read_only.cc",
+    "nested_namespace.cc",
+    "test.idl",
+    "v8tools_module.cc",
+    "xwalk_extensions_browsertest.cc",
+    "xwalk_extensions_test_base.cc",
+    "xwalk_extensions_test_base.h",
+  ]
+  defines = [ "HAS_OUT_OF_PROC_TEST_RUNNER" ]
+  deps = [
+    ":bad_extension",
+    ":bulk_data_transmission",
+    ":crash_extension",
+    ":echo_extension",
+    ":echo_extension_messaging_2",
+    ":generate_jsapi_extensions_test",
+    ":get_runtime_variable",
+    ":multiple_entry_points_extension",
+    "//base",
+    "//content/public/browser",
+    "//content/test:test_support",
+    "//net",
+    "//skia",
+    "//testing/gtest",
+    "//xwalk:xwalk_runtime",
+    "//xwalk/extensions",
+    "//xwalk/extensions:xwalk_extensions_resources",
+    "//xwalk/test/base:test_support",
+  ]
+  if (is_linux && !is_component_build && is_component_ffmpeg) {
+    configs += [ "//build/config/gcc:rpath_for_built_shared_libraries" ]
+  }
+}
+
+action("generate_jsapi_extensions_test") {
+  outputs = [
+    "$root_gen_dir/xwalk/extensions/test/test.h",
+    "$root_gen_dir/xwalk/extensions/test/test.cc",
+  ]
+  script = "//tools/json_schema_compiler/compiler.py"
+  inputs = [
+    "//tools/json_schema_compiler/cc_generator.py",
+    "//tools/json_schema_compiler/code.py",
+    "//tools/json_schema_compiler/compiler.py",
+    "//tools/json_schema_compiler/cpp_generator.py",
+    "//tools/json_schema_compiler/cpp_type_generator.py",
+    "//tools/json_schema_compiler/cpp_util.py",
+    "//tools/json_schema_compiler/h_generator.py",
+    "//tools/json_schema_compiler/idl_schema.py",
+    "//tools/json_schema_compiler/model.py",
+    "//tools/json_schema_compiler/util.cc",
+    "//tools/json_schema_compiler/util.h",
+    "//tools/json_schema_compiler/util_cc_helper.py",
+  ]
+  root_path = rebase_path("//xwalk/extensions", "$root_out_dir")
+  gen_path = rebase_path("$root_gen_dir", "$root_out_dir")
+  args = [
+    "$root_path/test/test.idl",
+    "--root=$root_path",
+    "--destdir=$gen_path/xwalk/extensions/",
+    "--namespace=xwalk::jsapi::test",
+    "--generator=cpp",
+  ]
+}
+
+shared_library("echo_extension") {
+  visibility = [ ":*" ]
+  sources = [
+    "echo_extension.c",
+  ]
+  output_dir = "$root_out_dir/tests/extension/echo_extension"
+}
+
+shared_library("echo_extension_messaging_2") {
+  visibility = [ ":*" ]
+  sources = [
+    "echo_extension_messaging_2.c",
+  ]
+  output_dir = "$root_out_dir/tests/extension/echo_extension"
+}
+
+shared_library("bad_extension") {
```

Comment:
We were talking about how `xwalk_extensions_browsertest` was broken in Debug mode earlier today because it was failing with errors such as

    error while loading shared libraries: libbad_extension.so: cannot open shared object file: No such file or directory

This come from the fact that you're using `shared_library` for these targets, which means the targets that _depend_ on them will link against them directly. In other words, linking the `xwalk_extensions_browsertest` binary will try to link all these extension .so's against it, which is definitely not what we want (the Release case works because GN passes `--as-needed` to the linker, which removes unneeded shared library dependencies automatically).

You actually need to change all these targets to `loadable_module` instead.